Introduce a `MessageGroup.condition` (#3517)
spring-projects/spring-integration: src/reference/asciidoc/message-store.adoc

This improves resources utilization when groups are very big in the store.
Internally in the framework this new API is used in the <<./delayer.adoc#delayer,Delayer>> (for example) when it reschedules persisted messages on startup.
A returned `Stream<Message<?>>` must be closed in the end of processing, e.g. via auto-close by the `try-with-resources`.
Whenever a `PersistentMessageGroup` is used, its `streamMessages()` delegates to the `MessageGroupStore.streamMessagesForGroup()`.

[[message-group-condition]]
==== Message Group Condition

Starting with version 5.5, the `MessageGroup` abstraction provides a `condition` string option.

> The value of this option can be anything what could be parsed later on for any reason to make a decision for the group.

Reviewer: ```suggestion
The value of this option can be anything that could be parsed later on for any reason to make a decision for the group.
```

For example a `ReleaseStrategy` from a <<./aggregator.adoc#aggregator-api, correlation message handler>> may consult this property from the group instead of iterating all the messages in the group.

> The `MessageGroupStore` expose a `setGroupCondition(Object groupId, String condition)` API.

Reviewer: ```suggestion
The `MessageGroupStore` exposes a `setGroupCondition(Object groupId, String condition)` API.
```

For this purpose a `setGroupConditionSupplier(BiFunction<Message<?>, String, String>)` option has been added to the `AbstractCorrelatingMessageHandler`.

> This function is evaluated against each message to be added to the group and existing condition on the group.

Reviewer: ```suggestion
This function is evaluated against each message after it has been added to the group as well as the existing condition of the group.
```

> The implementation may decide to return new value, existing or reset the target condition to `null`.

Reviewer: ```suggestion
The implementation may decide to return a new value, the existing value, or reset the target condition to `null`.
```

The value for a `condition` can be a JSON, SpEL expression, number or anything what can be serialized as a string and parsed afterwards.
For example, the `FileMarkerReleaseStrategy` from the <<./file.adoc#file-aggregator, File Aggregator>> component, populates a condition into a group from the `FileHeaders.LINE_COUNT` header of the `FileSplitter.FileMarker.Mark.END` message and consults with it from its `canRelease()` comparing a group size with the value in this condition.

> This way it doesn't iterate all the messages in group to find a `FileSplitter.FileMarker.Mark.END` message with the `FileHeaders.LINE_COUNT` header.

Reviewer: It also allows the end marker to arrive at the aggregator before all the other records; for example when processing a file in a multi-threaded environment.

> In addition, for target configuration convenience, a `GroupConditionProvider` contract has been introduced.
> The `AbstractCorrelatingMessageHandler` checks provided `ReleaseStrategy` for this contract implementation and extracts a `conditionSupplier` for further group condition evaluation logic.

Reviewer: ```suggestion
In addition, for configuration convenience, a `GroupConditionProvider` contract has been introduced.
The `AbstractCorrelatingMessageHandler` checks if the provided `ReleaseStrategy` implements this interface and extracts a `conditionSupplier` for group condition evaluation logic.
```
